### ISSUE TITLE
Tracking issue: Repo Management - Internal repo support

### ISSUE THREAD

**Issue by indradhanush** [sourcegraph/sourcegraph#32032]
## Plan

Support GitHub enterprise internal repos as a first class citizen. We have multiple issues related to permissions syncing of internal repos that although seem to appear to be related but are not. 

Each of the issues related to internal repos have a long discussion, so here's a list and a brief summary of each of them. 

<details>
    <summary>Click to expand</summary>

### [GitHub internal repositories are not synced when specifying "orgs" · Issue #17153 · sourcegraph/sourcegraph · GitHub](https://github.com/sourcegraph/sourcegraph/issues/17153)**

**Summary**

* Internal repositories not seen in site admin
* https://github.com/sourcegraph/sourcegraph/issues/17153#issuecomment-759512531
* With the following config:

```json
{
  "url": "https://ghe.sgdev.org",
  "token": "<REDACTED>",
  "authorization": {
    "ttl": "3h"
  },
  "orgs": [
    "sourcegraph"
  ]
}
```

> I'm able to both search/find the internal repository, and see it in the site-admin panel.
> \-@unknwon

* Token used in code host config should have access to the internal repo
* Users need to manually add an internal repo, doesn't show up automatically
* Doesn't show up when `orgs` is used, but shows up if specified explicitly under `repo` in code host config

### [authz/github: repo-centric groups sync does not account for repo being visibility `internal` · Issue #25904 · sourcegraph/sourcegraph · GitHub](https://github.com/sourcegraph/sourcegraph/issues/25904)

**Summary**

* Applicable when `groupsCacheTTL` enabled

## https://github.com/sourcegraph/sourcegraph/issues/21591

**Summary**

* **NOT related to #25904 IF `groupsCacheTTL` is disabled**
* GitHub sync is of two types:
  * Normal auth based sync
  * Cached sync that is when `groupsCacheTTL` is enabled

These seemingly related issues are a symptom of internal repos never being treated at the same level as public or private repos. It does not help that internal repos are a special case of private repos but for Github Enterprise Server instances.

</details>

### Tracked issues

- 👩 https://github.com/sourcegraph/sourcegraph/issues/17153
- 👩 https://github.com/sourcegraph/sourcegraph/issues/25904
- 👩 https://github.com/sourcegraph/sourcegraph/issues/21591
- 👩 https://github.com/sourcegraph/sourcegraph/issues/34684

### List of artifacts on existing work so far (to be kept updated as more artifacts are created)

- https://github.com/sourcegraph/sourcegraph/discussions/26837#discussioncomment-1611792
- https://github.com/sourcegraph/sourcegraph/pull/27272
- https://github.com/sourcegraph/sourcegraph/pull/29013
- https://github.com/sourcegraph/sourcegraph/pull/29325
- https://github.com/sourcegraph/sourcegraph/pull/30097
- https://github.com/sourcegraph/sourcegraph/pull/30296
- https://github.com/sourcegraph/sourcegraph/pull/30329
- https://github.com/sourcegraph/sourcegraph/pull/31261
- https://github.com/sourcegraph/sourcegraph/pull/31275
- https://github.com/sourcegraph/sourcegraph/pull/32025
- https://github.com/sourcegraph/sourcegraph/pull/32474
- https://github.com/sourcegraph/sourcegraph/pull/34696

### Plan of action

~Existing work has been focused on fixing #25904 and it has proven challenging (see list of artifacts above), however we are getting closer. With #25904 done, we will move our attention to #21591 because of customer prioritisation and then eventually fix #17153.~

**Update 2022-04-29**

Based on the workaround discovered by one of our customers, we will implement #34684 which should be able to resolve #25904 and #21591.

#### Legend

- 👩 Customer issue
- 🐛 Bug
- 🧶 Technical debt
- 🎩 Quality of life
- 🛠️ [Roadmap](https://handbook.sourcegraph.com/departments/product-engineering/process/planning-process#roadmap)
- 🕵️ [Spike](https://en.wikipedia.org/wiki/Spike_(software_development))
- 🔒 Security issue
- 🙆 Stretch goal

**Comment by indradhanush** [comment]
**Update 2022-03-29**

The artifact #32025 is now ready for review and brings us close with a moderately high degree of confidence to be able to complete the fix for #25904. 

The next step for this as soon as it is reviewed and merged is to test it in a customer-like setup and verify that permissions for internal repositories is syncing correctly. 

Post which we will move on to fixing #21591. While the fix for this won't be a trivial one either (unlike what I had anticipated based on my initial research last year), the good part is that I know why the bug exists and can start working on it without having to spend additional time in researching and validating it. 

cc: @dan-mckean, @ajaynz

**Comment by indradhanush** [comment]
**Update 2022-05-19**

We're halting work on this for the time being to regroup as a team on other work streams that has taken more priority for the moment. This remains in our radar but there's a low likelihood of it being picked up again in Q2 this year. 

cc: @ajaynz
